### PR TITLE
Avoid initializer name collision in _fuse_batchnorm.py

### PR DESCRIPTION
**Pull request by titaiwangms**
Fixes https://github.com/pytorch/pytorch/issues/166797

The original naming collides when there are multiple matched patterns sharing the same parent node. This PR changes the naming to depend on their own Conv weight name, which should be non-duplicated identifier.

~~NOTE: I don't know if my understanding is correct. It seems x is an input of the pattern, which x.name + "_bias" collides with `max_pool` bias (see the pic in the original issue)? If we check the output model after _fuse_batchnorm.py, the bias would be correct with a name `val_17` (the name may be collided and given by NameAuthority?). However, when the following rule _remove_optional_bias tries to fetch the bias, it would see all zero for some reasons.~~

### CODE REVIEW

File: onnxscript/rewriter/rules/common/_fuse_batchnorm.py [hunk 1]
```diff
@@ -68,7 +68,9 @@ def rewrite(self, op, x: ir.Value, inbound_out: ir.Value, batchnorm_out: ir.Valu
             bias_name = inbound_node.inputs[2].name
         else:
             original_bias = np.zeros_like(input_mean)
-            bias_name = x.name + "_bias"
```
Comment (justinchuby) [inline onnxscript/rewriter/rules/common/_fuse_batchnorm.py:71]: Curious why the original would collide?

Comment (titaiwangms) [reply onnxscript/rewriter/rules/common/_fuse_batchnorm.py:71]: It's hard to tell. Let me see if I can dig more.

Comment (AyoubMDL) [reply onnxscript/rewriter/rules/common/_fuse_batchnorm.py:71]: It is when there are multiple consumers. In the issue, there are two match patterns with the same parent, something like this 
image
I think it's good to take the weights name or the output name of the Conv

Comment (titaiwangms) [reply onnxscript/rewriter/rules/common/_fuse_batchnorm.py:71]: Thanks! I will add a test case.

Comment (titaiwangms) [reply onnxscript/rewriter/rules/common/_fuse_batchnorm.py:71]: Done

Comment (gramalingam) [reply onnxscript/rewriter/rules/common/_fuse_batchnorm.py:71]: Sounds like:
1. When we create initializers, we need a way to create new unique names (at least unique among initializers, to begin with). I wonder if we can expose an IR method that will change an initializer's name to be unique (instead of throwing an exception) when adding it?
2. In this example, it should work fine since the first inputs are different. What happens if the first inputs are the same in two nodes? Granted, it won't happen in realistic examples. But it could still cause an exception, I think?

Comment (titaiwangms) [reply onnxscript/rewriter/rules/common/_fuse_batchnorm.py:71]: > When we create initializers, we need a way to create new unique names (at least unique among initializers, to begin with). I wonder if we can expose an IR method that will change an initializer's name to be unique (instead of throwing an exception) when adding it?

Agree. In the created case, from what I saw, we don't even see an exception, the fused bias from pattern 1 and 2 co-exists with the same names.

> 2. In this example, it should work fine since the first inputs are different. What happens if the first inputs are the same in two nodes? Granted, it won't happen in realistic examples. But it could still cause an exception, I think?

In terms of Conv node, the example is problematic when bias is fused from op.BatchNorm to op.Conv because they will share the same naming.

**Review by justinchuby — APPROVED** [review]

File: onnxscript/rewriter/rules/common/_fuse_batchnorm.py (hunk 1 above)
Comment (titaiwangms) [reply onnxscript/rewriter/rules/common/_fuse_batchnorm.py:71]: Merge for now

Comment (titaiwangms) [reply onnxscript/rewriter/rules/common/_fuse_batchnorm.py:71]: Feel free to continue the discussion

File: onnxscript/rewriter/rules/common/_fuse_batchnorm.py [hunk 2]
```diff
@@ -68,7 +68,10 @@ def rewrite(self, op, x: ir.Value, inbound_out: ir.Value, batchnorm_out: ir.Valu
...
+            # Use inbound input 1 (should be weight) to derive a name for the bias
+            # to avoid name collision on initializer creation when there are multiple patterns
+            # sharing the same parent nodes.
+            bias_name = inbound_node.inputs[1].name + "_bias"
```
Comment (xadupre) [inline onnxscript/rewriter/rules/common/_fuse_batchnorm.py:74]: add one output name into the name, that way it will be unique

File: onnxscript/rewriter/rules/common/_fuse_batchnorm.py (hunk 1 above)
Comment (titaiwangms) [reply onnxscript/rewriter/rules/common/_fuse_batchnorm.py:71]: https://github.com/onnx/ir-py/issues/246
